include/functions.php

```diff
@@ -574,6 +574,27 @@ function generate_page_title($page_title, $p = null)
 }
 
 
+//
+// Generate breadcrumb navigation
+//
+function generate_crumbs($crumbs)
+{
+	$last_key = end(array_keys($crumbs));
+
+	$crumbs_markup = '<ol class="crumbs">';
+	foreach ($crumbs as $idx => $crumb)
+	{
+		$crumbs_markup .= '<li>'.($idx ? '<span>»&#160;</span>' : '');
+		$item = is_array($crumb) ? '<a href="'.$crumb[1].'">'.pun_htmlspecialchars($crumb[0]).'</a>' : $crumb;
+		$crumbs_markup .= $idx == $last_key ? '<strong>'.$item.'</strong>' : $item;
```

Comment:
For consistency, I'd like to see `$crumb` (one line above) and `$item` be escaped, too. Probably not necessary for the current usages, but again safer than having to remember to add it when it becomes necessary.

Reply:
Good catch! Here is another way to do this. Do you have a preference?

```
function generate_crumbs($crumbs)
{
	$crumbs_markup = '<ol class="crumbs">';

	$num_crumbs = count($crumbs);
	for ($i = 0; $i < $num_crumbs; ++$i)
	{
		$crumbs_markup .= '<li>'.(($i) ? '<span>»&#160;</span>' : '');
		$item = (is_array($crumbs[$i])) ? '<a href="'.$crumbs[$i][1].'">'.pun_htmlspecialchars($crumbs[$i][0]).'</a>' : pun_htmlspecialchars($crumbs[$i]);
		$crumbs_markup .= ($i == $num_crumbs - 1) ? '<strong>'.$item.'</strong>' : $item;
		$crumbs_markup .= '</li>';
	}

	$crumbs_markup .= '</ol>';

	return $crumbs_markup;
}
```

Reply:
No strong preference. I find the original version a little easier to read.

You could even *assume* there are at least two elements in `$crumbs` and handle the two special cases (first and last iteration) outside of the loop...

Reply:
When `pun_htmlspecialchars` applied to the last `$crumb`:
```
    Index » Search » <a href="search.php?action=show_recent">Active</a>
```

Reply:
You'll have to unpack the `$crumbs_text` stuff in search.php. All that is doing is wrapping text in a link, as far as I can tell - and you can pass these two in an array to your `generate_crumbs()` function.

Reply:
Can you show the code for the first and last iteration?

Reply:
I got it to work except for the following:

Original code:
```
$crumbs_text['search_type'] = '<a href="search.php?action=search&amp;keywords='.urlencode($keywords).'&amp;author='.urlencode($author).'&amp;forums='.$search_type[2].'&amp;search_in='.$search_type[3].'&amp;sort_by='.$sort_by.'&amp;sort_dir='.$sort_dir.'&amp;show_as='.$show_as.'">'.$crumbs_text['search_type'].'</a>';
```
Revised code:
```
$crumbs_text['search_type'] = array($crumbs_text['search_type'], 'search.php?action=search&amp;keywords='.urlencode($keywords).'&amp;author='.urlencode($author).'&amp;forums='.$search_type[2].'&amp;search_in='.$search_type[3].'&amp;sort_by='.$sort_by.'&amp;sort_dir='.$sort_dir.'&amp;show_as='.$show_as);
```
Output when the keyword search is `<strong>`:
```
Index » Search » Topics with posts containing '&lt;strong&gt;'
```
Markup:
```
<ol class="crumbs"><li><a href="index.php">Index</a></li><li><span>»&#160;</span><a href="search.php">Search</a></li><li><span>»&#160;</span><strong><a href="search.php?action=search&amp;keywords=%3Cstrong%3E&amp;author=&amp;forums=&amp;search_in=0&amp;sort_by=0&amp;sort_dir=DESC&amp;show_as=topics">Topics with posts containing &#039;&amp;lt;strong&amp;gt;&#039;</a></strong></li></ol>
```

Reply:
We may have to revert `pun_htmlspecialchars` to have more flexibility. If yes, is there a way to revert the changes from yesterday?

Reply:
You can [undo commits with Git](https://stackoverflow.com/questions/927358/how-do-i-undo-the-most-recent-commits-in-git), and then force-push.

I can have a look again next week, will be AFK for a few days.